### ISSUE TITLE
Load a TF model from memory, not file (C/C++)

### ISSUE THREAD

**Issue by MaartenSneepKNMI** [tensorflow/tensorflow#37570]
**System information**
- TensorFlow version (you are using): 1.12.0
- Are you willing to contribute it (Yes/No): Yes, if possible.
- C/C++ interface.

**Describe the feature and the current behavior/state.**

We'd like to load model data from memory without file access.

We're using TF in a satellite data processing system. One of the requirements dictates that file access is not allowed, we can only load from memory. Of course the model is stored somewhere, but that is very much outside the parts we can reach from where we run, we can only receive the bytestream of that file in a memory block.

To meet this requirement, we privately patched version 1.12.0 of TF. Because we like living on the edge, we're also using the C++ (unstable) interface, so we're tied rather hard to this version at the moment. We realise that implementation of this feature request will require changes to our code, but will also make it a lot easier to move to the current version, especially now that the C interface seems to be stable (although statements in different parts on the site are inconsistent).

**Will this change the current api? How?**

This would be an _addition_ to the current interface. I don't see how the external interfaces would need to change, although internally the file loaders may want to use this new interface. 

**Who will benefit with this feature?**

Besides us, this may also be useful for the Java interface, as apache interfaces usually don't like files either. 

**Any Other info.**

None.

**Comment by ibrahimsoliman97** [comment]
Any update about this feature ?

**Comment by MaartenSneepKNMI** [comment]
This is the first response I received.

Best,

Maarten

This message is short because it was sent from my iPhone. Everything is well with me otherwise.

Op 10 aug. 2020 om 11:27 heeft Ibrahim Soliman <notifications@github.com> het volgende geschreven:

﻿

Any update about this feature ?

**Comment by rmarru** [comment]
Considering the changes introduces in tensorflow 2.0.0, specially focusing in `SavedModel` format, I don't think google/tensorflow team has any intention to go towards this direction but the opposite one.

From my experience, it is possible to create a model with tensorflow 2.x, save it using the `compat.v1` API - `tf.compat.v1.saved_model.save` - and convert it to a protobuf using the `freeze_model` scripts. Then, you'll be able to load this protobuf into memory while using the existing tensorflow C API to load a model from a protobuf.

**Comment by MaartenSneepKNMI** [comment]
> Considering the changes introduced in tensorflow 2.0.0, specially focusing in `SavedModel` format, I don't think google/tensorflow team has any intention to go towards this direction but the opposite one.

This is unfortunate, as EUMETSAT is moving in a direction where direct file access is prohibited. 

> From my experience, it is possible to create a model with tensorflow 2.x, save it using the `compat.v1` API - `tf.compat.v1.saved_model.save` - and convert it to a protobuf using the `freeze_model` scripts. Then, you'll be able to load this protobuf into memory while using the existing tensorflow C API to load a model from a protobuf.

Forgive me, I may be a little dense here. Are you referring to the C API in version 2 here? As in: using the C API may be a way out to using a stock library, using a stable API, and still meet the requirements set by EUMETSAT?

**Comment by rmarru** [comment]
> Forgive me, I may be a little dense here. Are you referring to the C API in version 2 here? As in: using the C API may be a way out to using a stock library, using a stable API, and still meet the requirements set by EUMETSAT?

Yes, out team managed to do so with the Tensorflow 2.1.0 C API. Concretely, you can find functions as [TF_GraphImportGraphDef](https://github.com/tensorflow/tensorflow/blob/fedc6d951faa73936a1154d6507d54240614d416/tensorflow/c/c_api.cc#L1804) in order to load a graph from a graph_def (your protobuf).

**Comment by MaartenSneepKNMI** [comment]
Thank you, this opens a way out for us.

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.

**Comment by DanijelDomazet** [comment]
Any news on loading a model from memory instead of a file?

**Comment by MaartenSneepKNMI** [comment]
The problem has been resolved, but not in a way that is useful to you, probably.

1. We asked to remove the requirement that no file access was performed, at least for the initialisation phase of our processor. This has been granted, and this obviates the need for this change for us.
2. Perhaps more importantly, we dropped TensorFlow for _using_ the neural network. We still use TF to generate the NN (Keras interface in Python), but for our use case, it turned out to be a lot easier to code the matrix multiplications and other manipulations ourselves rather than carrying the full bagage of the library with us.
